packages/arb-token-bridge-ui/src/hooks/useTransactionHistory.ts

```diff
@@ -332,35 +337,138 @@ export const useTransactionHistory = (
-  >(address ? ['new_tx_list', address] : null)
+  const { data: newTransactionsData, mutate: mutateNewTransactionsData } =
+    useSWRImmutable<MergedTransaction[]>(
+      address ? ['new_tx_list', address] : null
+    )
+
+  const transactions: MergedTransaction[] = [
+    ...(newTransactionsData || []),
+    ...(txPages || [])
+  ].flat()
 
   const addPendingTransaction = useCallback(
     (tx: MergedTransaction) => {
       if (!isTxPending(tx)) {
         return
       }
 
-      mutate(currentNewTransactions => {
+      mutateNewTransactionsData(currentNewTransactions => {
         if (!currentNewTransactions) {
           return [tx]
         }
 
         return [tx, ...currentNewTransactions]
       })
     },
-    [mutate]
+    [mutateNewTransactionsData]
+  )
+
+  const updateCachedTransaction = useCallback(
+    (newTx: MergedTransaction) => {
+      // check if tx is a new transaction initiated by the user, and update it
+      const foundInNewTransactionsCache = !!newTransactionsData?.find(oldTx =>
+        isSameTransaction(oldTx, newTx)
+      )
+
+      if (foundInNewTransactionsCache) {
+        // replace the existing tx with the new tx
+        mutateNewTransactionsData(txs =>
+          txs?.map(oldTx => {
+            return { ...(isSameTransaction(oldTx, newTx) ? newTx : oldTx) }
+          })
+        )
+        return
+      }
+
+      // tx not found in the new user initiated transaction list
+      // look in the paginated historical data
+      if (!txPages) {
+        return
+      }
+
+      let pageNumberToUpdate = 0
+
+      // search cache for the tx to update
+      while (
+        !txPages[pageNumberToUpdate]?.find(oldTx =>
+          isSameTransaction(oldTx, newTx)
+        )
+      ) {
+        pageNumberToUpdate++
+
+        if (pageNumberToUpdate > txPages.length) {
+          // tx not found
+          return
+        }
+      }
+
+      const oldPageToUpdate = txPages[pageNumberToUpdate]
+
+      if (!oldPageToUpdate) {
+        return
+      }
+
+      // replace the old tx with the new tx
+      const updatedPage = oldPageToUpdate.map(oldTx => {
+        return isSameTransaction(oldTx, newTx) ? newTx : oldTx
+      })
+
+      // all old pages including the new updated page
+      const newTxPages = [
+        ...[...txPages].slice(0, pageNumberToUpdate),
+        updatedPage,
+        ...[...txPages].slice(pageNumberToUpdate + 1)
+      ]
+
+      mutateTxPages(newTxPages, false)
+    },
+    [mutateNewTransactionsData, mutateTxPages, newTransactionsData, txPages]
+  )
+
+  const updatePendingTransaction = useCallback(
+    async (tx: MergedTransaction) => {
+      const foundInCache = !!transactions.find(t => isSameTransaction(t, tx))
+
+      if (!foundInCache) {
+        // tx does not exist
+        return
+      }
+
+      if (!isTxPending(tx)) {
+        // if not pending we don't need to check for status, we accept whatever status is passed in
+        updateCachedTransaction(tx)
```

Comment:
If the tx is not pending do we have to do anything with it? Can we just return?

Reply:
We still have cases where we want to update a transaction if the new status isn't pending.

For example, if we claim and the claim is successful, the new status will be success so we don't have to run heavy fetching of the status again. In that case we can directly mutate the transaction.